src/sentry/web/frontend/auth_login.py

```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
+            ),
+            limit=5,
+            window=60,  # 5 per minute should be enough for anyone
+        )
+
+    def get_ratelimited_login_form(
+        self, request: Request, login_form: AuthenticationForm, op: str, **kwargs
+    ) -> HttpResponse:
+        """Returns a login form with ratelimited errors displayed."""
+        login_form.errors["__all__"] = [
+            "You have made too many login attempts. Please try again later."
+        ]
+        metrics.incr("login.attempt", instance="rate_limited", skip_internal=True, sample_rate=1.0)
+
+        context = {
+            "op": op or "login",
+            "login_form": login_form,
+        }
+
+        context.update(additional_context.run_callbacks(request))
+        return self.respond_login(request=request, context=context, **kwargs)
+
+    def redirect_to_next_login_step(
+        self,
+        request: Request,
+        login_form: AuthenticationForm,
+        organization: Organization,
+    ) -> HttpResponseRedirect:
+        """Called when a user submits a valid login form and must be redirected to
+        the next step in their auth process"""
+        user = login_form.get_user()
+
+        self._handle_login(request=request, user=user, organization=organization)
+        metrics.incr("login.attempt", instance="success", skip_internal=True, sample_rate=1.0)
+
+        if not user.is_active:
+            return self.redirect(url=reverse("sentry-reactivate-account"))
+        if organization:
+            self.refresh_organization_status(request=request, user=user, organization=organization)
+        # On login, redirect to onboarding
+        if self.active_organization:
+            if features.has(
+                "organizations:customer-domains",
+                self.active_organization.organization,
+                actor=user,
+            ):
+                setattr(request, "subdomain", self.active_organization.organization.slug)
+        return self.redirect(url=get_login_redirect(request=request))
+
+    def _handle_login(
+        self, request: Request, user: User, organization: Optional[Organization]
+    ) -> None:
+        """Logs a user in and determines their active org."""
+        login(request=request, user=user, organization_id=coerce_id_from(m=organization))
+        self.determine_active_organization(request=request)
+
+    def refresh_organization_status(
+        self, request: Request, user: User, organization: Organization
+    ) -> None:
+        """Refresh organization status/context after a successful login to inform other interactions."""
+        # Refresh the organization we fetched prior to login in order to check its login state.
+        org_context = organization_service.get_organization_by_slug(
+            user_id=request.user.id,
+            slug=organization.slug,
+            only_visible=False,
+        )
+        if org_context:
```

Comment:
can we add some notes to what we're doing with the context here?

Looks like we're setting the active org on the session and/or stripping the 'next' redirect if the user is not an org member?

Reply:
yea ill need to follow up with some more people to properly understand what this is doing, then ill document it better.